scikeras/_saving_utils.py

```diff
@@ -0,0 +1,140 @@
+
+
+def _restore_optimizer_weights(optimizer, weights) -> None:
+    optimizer._restored_weights = weights
+    optimizer._create_all_weights_orig = optimizer._create_all_weights
+    # MethodType is used to "bind" the _temp_create_all_weights method
+    # to the "live" optimizer object
+    optimizer._create_all_weights = MethodType(_temp_create_all_weights, optimizer)
+
+
+def unpack_keras_model(packed_keras_model, optimizer_weights):
+    """Reconstruct a model from the result of __reduce__
+    """
+    temp_dir = _get_temp_folder()
+    b = BytesIO(packed_keras_model)
+    with zipfile.ZipFile(b, "r", zipfile.ZIP_DEFLATED) as zf:
+        for path in zf.namelist():
+            dest = os.path.join(temp_dir, path)
+            tf_io.gfile.makedirs(os.path.dirname(dest))
+            with tf_io.gfile.GFile(dest, "wb") as f:
+                f.write(zf.read(path))
+    model: keras.Model = load_model(temp_dir)
+    for root, _, filenames in tf_io.gfile.walk(temp_dir):
+        for filename in filenames:
+            if filename.startswith("ram://"):
+                # Currently, tf.io.gfile.walk returns
+                # the entire path for the ram:// filesystem
+                dest = filename
+            else:
+                dest = os.path.join(root, filename)
+            tf_io.gfile.remove(dest)
+    _restore_optimizer_weights(model.optimizer, optimizer_weights)
+    return model
+
+
+def pack_keras_model(model):
+    """Support for Pythons's Pickle protocol.
+    """
+    temp_dir = _get_temp_folder()
+    model.save(temp_dir)
+    b = BytesIO()
+    with zipfile.ZipFile(b, "w", zipfile.ZIP_DEFLATED) as zf:
+        for root, _, filenames in tf_io.gfile.walk(temp_dir):
+            for filename in filenames:
+                if filename.startswith("ram://"):
+                    # Currently, tf.io.gfile.walk returns
+                    # the entire path for the ram:// filesystem
+                    dest = filename
+                else:
+                    dest = os.path.join(root, filename)
+                with tf_io.gfile.GFile(dest, "rb") as f:
+                    zf.writestr(os.path.relpath(dest, temp_dir), f.read())
+                tf_io.gfile.remove(dest)
+    b.seek(0)
+    return (
+        unpack_keras_model,
+        (np.asarray(memoryview(b.read())), model.optimizer.get_weights()),
+    )
+
+
+def unpack_keras_optimizer(opt_serialized, weights):
+    """Reconstruct optimizer.
+    """
+    optimizer: keras.optimizers.Optimizer = keras.optimizers.deserialize(opt_serialized)
+    _restore_optimizer_weights(optimizer, weights)
+    return optimizer
+
+
+def pack_keras_optimizer(optimizer: keras.optimizers.Optimizer):
+    """Support for Pythons's Pickle protocol in Keras Optimizers.
+    """
+    opt_serialized = keras.optimizers.serialize(optimizer)
+    weights = optimizer.get_weights()
+    return unpack_keras_optimizer, (opt_serialized, weights)
+
+
+def unpack_keras_metric(metric_serialized):
```

Comment:
These functions for metric/loss don't do much.

Reply:
They are very simple, yes, but still required to have a single function we can attach as `__reduce__`. I guess one alternative, that also applies to the other monkey patching, would be to do:

```python
from tf.keras import metrics

class Metric(metrics.Metric):
    def __reduce__(...):
       ...
metrics.Metric = Metric
```

I guess this could work, but I'm not sure that it's cleaner.

Reply:
Does that example change the user's version of `tf.keras.Metric`? Should a user ever import it? Or is it only used in SciKeras?

If a user passes a custom metric, can that metric be serialized?

Reply:
We need to change the global version of `Metric` in order for all metrics (including custom user metrics) to be serializable. I don't see any downside to this as it doesn't break anything as far as I can tell.

The alternative to modifying the global/user version would be to have a hook in `__init__` much like we have now that monkey patches the method onto live instances of `Model`, `Optimizer`, etc. instead of onto the classes like I am doing here. The con of this is that it (1) requires more code and (2) means that if a user were to define a custom parameter and passed an `Optimizer` to it (something like `KerasClassifier(..., my_optimizers=[Adam(), SGD()])` they'd get some cryptic error about `RLock` objects or they might not even get an error and instead get unpredictable results